cpp/src/arrow/flight/sql/server_session_middleware.cc

```diff
@@ -0,0 +1,179 @@
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "arrow/flight/sql/server_session_middleware.h"
+#include <boost/lexical_cast.hpp>
+#include <boost/uuid/uuid.hpp>
+#include <boost/uuid/uuid_generators.hpp>
+#include <boost/uuid/uuid_io.hpp>
+
+namespace arrow {
+namespace flight {
+namespace sql {
+
+/// \brief A factory for ServerSessionMiddleware, itself storing session data.
+class ServerSessionMiddlewareFactory : public ServerMiddlewareFactory {
+ protected:
+  std::map<std::string, std::shared_ptr<FlightSqlSession>> session_store_;
+  std::shared_mutex session_store_lock_;
+  boost::uuids::random_generator uuid_generator_;
+
+  std::vector<std::pair<std::string, std::string>> ParseCookieString(
+      const std::string_view& s) {
+    const std::string list_sep = "; ";
+    const std::string pair_sep = "=";
+    const size_t pair_sep_len = pair_sep.length();
+
+    std::vector<std::pair<std::string, std::string>> result;
+
+    size_t cur = 0;
+    while (cur < s.length()) {
+      const size_t end = s.find(list_sep, cur);
+      size_t len;
+      if (end == std::string::npos) {
+        // No (further) list delimiters
+        len = std::string::npos;
+        cur = s.length();
+      } else {
+        len = end - cur;
+        cur = end;
+      }
+      const std::string_view tok = s.substr(cur, len);
+
+      const size_t val_pos = tok.find(pair_sep);
+      result.emplace_back(tok.substr(0, val_pos),
+                          tok.substr(val_pos + pair_sep_len, std::string::npos));
+    }
+
+    return result;
+  }
+
+ public:
+  Status StartCall(const CallInfo&, const CallHeaders& incoming_headers,
+                   std::shared_ptr<ServerMiddleware>* middleware) {
+    std::string session_id;
+
+    const std::pair<CallHeaders::const_iterator, CallHeaders::const_iterator>&
+        headers_it_pr = incoming_headers.equal_range("cookie");
+    for (auto itr = headers_it_pr.first; itr != headers_it_pr.second; ++itr) {
+      const std::string_view& cookie_header = itr->second;
+      const std::vector<std::pair<std::string, std::string>> cookies =
+          ParseCookieString(cookie_header);
+      for (const std::pair<std::string, std::string>& cookie : cookies) {
+        if (cookie.first == kSessionCookieName) {
+          session_id = cookie.second;
+          if (session_id.empty())
+            return Status::Invalid("Empty " +
+                                   static_cast<std::string>(kSessionCookieName) +
+                                   " cookie value.");
```

Comment:
```suggestion
            return Status::Invalid("Empty ", kSessionCookieName, " cookie value");
```

Reply:
Not valid though: both are const char[].  Probably more relevant to static_cast the first literal though, what I'm doing is a bit weird too (fixed build error late at night and didn't think very hard about the types other than that of kSessionCookieName).

Reply:
Are you sure this doesn't work? The constructor here uses a variadic template called StringBuilder that should handle things like this.

Reply:
Sorry, yes it works as variadic args; I misread as modifications to operator+ usage.